unique-universes/src/views/shoot.py

```diff
@@ -0,0 +1,295 @@
+from __future__ import annotations
+
+import dataclasses
+import math
+import random
+from asyncio import sleep
+from typing import TYPE_CHECKING, cast
+
+import disnake
+
+# weird import but it's to avoid circular imports
+import src.exts.minigames as games
+from src.bot import Universe
+from src.database import PlayerExistsError
+
+if TYPE_CHECKING:
+    from src.database import PlayerData
+
+
+# acceleration directed to the center of Earth
+# measured in m/s^2
+EARTH_ACCELERATION = 9.81
+# same thing, just for the moon
+MOON_ACCELERATION = 1.62
+
+
+class AngleModal(disnake.ui.Modal):
+    def __init__(self, view: ShootMenu) -> None:
+        self.view = view
+
+        super().__init__(
+            title="Angle Input",
+            components=[
+                disnake.ui.TextInput(
+                    label="Angle degrees",
+                    custom_id="angle_deg",
+                    style=disnake.TextInputStyle.short,
+                    placeholder="15",
+                    value=str(self.view.stats.angle),
+                    min_length=1,
+                    max_length=3,
+                ),
+            ],
+            timeout=180,
+        )
+
+    async def on_timeout(self) -> None:
+        await self.view.on_timeout()
+        self.view.stop()
+
+    async def callback(self, interaction: disnake.ModalInteraction) -> None:
+        angle = interaction.text_values["angle_deg"]
+        if not angle.isdigit():
```

Comment:
`.isdigit` accepts a lot more than simple numbers. The most sure-fire way to check if a string represents a number is to try casting it to int and catch the error if it's not.